### PR TITLE
Add two tier caching

### PR DESCRIPTION
**Pull request by hurshagrawal**
Adds 2-tier caching from https://github.com/mozilla/sccache/pull/241/files to our fork.

Seems to work (although still getting a lot of cache misses):
image

sccache logs:
image

### CODE REVIEW

**Comment by hurshagrawal** [comment]
> 1. What happens if there's a cache miss on s3?

It compiles and fills the disk cache.

> 2. Does it fill the cache if there's a hit from the disk?

It fills the disk cache only. Doesn't fill the remote cache at all. In the original Electron implementation, only CI updates the remote cache.

> 2. Does it fill both caches if there's a cache miss everywhere?

No, just disk cache.

Honestly I got about 1500 hits out of the 16k so not sure this was totally worth it but I guess it'll speed up first builds for people slightly.

**Comment by iliabylich** [comment]
> Honestly I got about 1500 hits out of the 16k so not sure this was totally worth it but I guess it'll speed up first builds for people slightly.

Interesting. Which version of electron/chromium did you compile? Electron `7-1-x` and chromium that is specified in its `DEPS` file? When I tested it I had 99% cache hit. Maybe the logic that it uses to count cache hits is broken? (i.e. it may count local cache miss followed by s3 cache hit as a miss+hit)

**Comment by hurshagrawal** [comment]
> Interesting. Which version of electron/chromium did you compile? Electron 7-1-x and chromium that is specified in its DEPS file? When I tested it I had 99% cache hit. Maybe the logic that it uses to count cache hits is broken? (i.e. it may count local cache miss followed by s3 cache hit as a miss+hit)

Yep the one in the DEPS file. So weird! My architecture isn't any different either (Macbook Pro). Not really sure then... could be counting or cache key issues?

File: src/cache/two_tier_disk.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                            as Box<dyn futures::Future<Error=errors::Error, Item=Cache>>;
+                    }
+
+                    Box::new(remote.get(&key)
+                        .then(move |remote_status| {
+                            match remote_status {
+                                Ok(Cache::Hit(mut entry)) => {
+                                    {
+                                        trace!("cache hit but need to push back into primary cache");
+                                        disk.put(&key, entry.to_write()).wait();
```
Comment (iliabylich) [inline src/cache/two_tier_disk.rs]: ideally we should do `remote.put` here as well

Comment (hurshagrawal) [reply src/cache/two_tier_disk.rs]: Ok lemme try!

Comment (hurshagrawal) [reply src/cache/two_tier_disk.rs]: Hm actually this means for every local disk cache hit we'll also make an http request to s3 to see if it exists there before moving forward to the next target, which may slow local compilation down (and make compilation require network connection). Let's keep as is for now?

Comment (hurshagrawal) [reply src/cache/two_tier_disk.rs]: I'll take another look if it continues getting a lot of cache misses.

**Pull request merged**